### PR TITLE
[BEAM-59] Beam FileSystem: initial set of methods and their implementation for local and gcs.

### PR DESCRIPTION
**Pull request by peihe**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by peihe** [comment]
R: @dhalperi 

match() and resolve() will be sent in separate PRs.

**Comment by peihe** [comment]
retest this please

**Comment by peihe** [comment]
PTAL
ready for review

**Comment by peihe** [comment]
PTAL

**Comment by dhalperi** [comment]
We had discussed the need for test coverage of all functions. Please add the missing tests before I TAL.

**Comment by peihe** [comment]
@dhalperi 
I thought we agreed that you can start the review, while I am improve the test coverage.
why tests coverage block you from review at least the interface and javadoc?

and, the implementation and tests is a line-by-line straight port from FileIOChannelFacotry to LocalFileSystem. It should be trivial to review.
And, I agree more test coverage is better. But, what is blocking us from merge this PR, and adding tests later?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystem.java [hunk 1]
```diff
@@ -19,16 +19,65 @@
...
+import java.util.List;
 import org.apache.beam.sdk.io.FileSystem;
+import org.apache.beam.sdk.io.fs.CreateOptions;
 import org.apache.beam.sdk.options.GcsOptions;
 
 /**
  * {@link FileSystem} implementation for Google Cloud Storage.
  */
-class GcsFileSystem extends FileSystem {
+public class GcsFileSystem extends FileSystem<GcsResourceId> {
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystem.java]: Why is this public?

**Comment by dhalperi** [comment]
LGTM as a faithful copy and paste.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystem.java (hunk 1 above)
Comment (peihe) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystem.java]: reverted

**Pull request closed without merging**